ipol_demo/modules/demorunner/demorunner.py

```diff
@@ -28,847 +28,713 @@
+                            f"Construct failed: unauthorized access. Move: {file_to_move}"
+                        )
+
+                    # Check origin
+                    if not os.path.exists(path_from):
+                        raise IPOLCompileError(
+                            f"Construct failed. File not found. Construct can't move file since it doesn't exist: {path_from}"
+                        )
+
+                    try:
+                        # Remove path_to if it exists
+                        remove_path(path_to)
+                        # Do move
+                        shutil.move(path_from, path_to)
+                    except (IOError, OSError):
+                        logger.error(
+                            f"construct: Can't move file {path_from} --> {path_to}",
+                        )
+                        # If can't move, write in the log file, so
+                        # the user can see it
+                        f = open(log_file, "w")
+                        f.write(f"Failed to move {path_from} --> {path_to}")
+                        f.close()
+                        raise
+    finally:
+        release_lock(lock_path)
+
+
+def construct_is_locked(lock_path):
+    """
+    Check if compilation path is locked
+    """
+    if os.path.isfile(lock_path):
+        current_time = time.time()
+        creation_time = os.path.getctime(lock_path)
+        # In case of error it might leave dangling lockfiles, remove it if it's old.
+        if current_time - creation_time >= 20 * 60:
+            release_lock(lock_path)
+            return False
+        return True
+    return False
+
+
+def acquire_construct_lock(lock_path):
+    """
+    Create compilation path lock in order to protect against simultaneous compilations
+    """
+    open(lock_path, "w+").close()
+
+
+def release_lock(lock_path):
+    """
+    Releases (removes) a lock given its path
+    """
+    try:
+        os.remove(lock_path)
+    except FileNotFoundError:
+        pass
+
+
+def any_extraction_needed(ddl_builds, dl_dir):
+    """
+    Check if binaries should be extracted
+    """
+    extract_needed = False
+    for build_item in list(ddl_builds.values()):
+        url = build_item.get("url")
+        if not url:
+            raise IPOLCompileError(f"Missing build item: url. ddl_build: {ddl_builds}")
+        username = build_item.get("username")
+        password = build_item.get("password")
+        zip_filename = urllib.parse.urlsplit(url).path.split("/")[-1]
+        tgz_file = os.path.join(dl_dir, zip_filename)
+        if build.download(url, tgz_file, username, password):
+            extract_needed = True
+    return extract_needed
+
+
+def all_files_exist(files):
+    """
+    Checks if all given file names exist
+    """
+    return all([os.path.isfile(f) or os.path.isdir(f) for f in files])
+
+
+@app.post("/compilations/{demo_id}", status_code=201)
+def ensure_compilation(demo_id: int, build_data: dict) -> None:
```

Comment:
replace `build_data` with `ddl_build` directly? and optional ssh public/private keys